# Assignment 9.2

> Replace all TODOs with your code. Do not change any other code.

In [1]:
# Do not edit this cell

from typing import List

## Descriptive statistics

In this assignment, we will write the functions to calculate the basic statistics from scratch, not using numpy.

### Task 1

Let's start simple: write a function `mean` that calculates the average of the list.

$$\mu = \frac{{\sum_{i=1}^n x_i}}{{n}}$$

In [3]:
def mean(li: List[float]) -> float:
    return sum(li) / len(li)

assert mean([1., 2., 3.]) == 2.
assert mean([1., 1., 2., 0.]) == 1.

### Task 2

Now let's calculate variance (dispersion). You may use the `mean` function implemented before.

$$V = \frac{{\sum_{i=1}^n (x_i - \mu)^2}}{{n}}$$

In [6]:
def variance(li: List[float]) -> float:
    μ = mean(li)
    return sum((x - μ) ** 2 for x in li) / len(li)


assert variance([1., 1., 1.]) == 0.
assert variance([1., 2., 3., 4.]) == 1.25

### Task 3

The standard deviation is easy once you get the variance:

$$\sigma = \sqrt{V}$$

In [7]:
def std(li: List[float]) -> float:
    return variance(li) ** 0.5

assert std([1., 1., 1.]) == 0.
assert std([1., 2., 3., 4.]) == 1.25**0.5

### Task 4

**Median**

The median is the middle value in a sorted dataset. If the dataset has an odd number of values, the median is the value at the center. If the dataset has an even number of values, the median is the average of the two middle values.

In [8]:
def median(li: List[float]) -> float:
    li = sorted(li)
    n = len(li)
    mid = n // 2
    return (li[mid] if n % 2 != 0 else (li[mid - 1] + li[mid]) / 2)

assert median([1., 1., 1.]) == 1.
assert median([1., 4., 3., 2.]) == 2.5

## Measure performance

Sometimes, apart from theoretical, algorithmic complexity, it's a good idea to compare the runtime of two algorithms empirically, i.e., run the code many times and time it.

In Python's standard library, we have [timeit](https://docs.python.org/3/library/timeit.html) module that does exactly that.

Let's compare the runtime of your implementations and numpy. Use the provided setup code:

In [9]:
import timeit

# generate data for tests
setup = '''
import random
import numpy as np
from __main__ import mean, variance, std, median

arr = np.random.rand(10_000) * 100
li = [random.random() * 100 for _ in range(10_000)]
'''

# pass your function to timeit module
funcs = {
    'mean': mean,
    'variance': variance,
    'std': std,
    'median': median,
}

for name, func in funcs.items():
    time = timeit.timeit(f"{func.__name__}(li)", setup=setup, number=100)
    print(f"{name}: {time:.5f} sec")


mean: 0.01109 sec
variance: 0.21469 sec
std: 0.18083 sec
median: 0.17100 sec


### Task 5

Complete Python statements to compare your functions to numpy. Use `li` for your function and `arr` for numpy functions.

In [10]:
stmt_mean_custom = 'mean(li)'
stmt_mean_np = 'np.mean(arr)'

stmt_var_custom = 'variance(li)'
stmt_var_np = 'np.var(arr)'

stmt_std_custom = 'std(li)'
stmt_std_np = 'np.std(arr)'

stmt_median_custom = 'median(li)'
stmt_median_np = 'np.median(arr)'

### Task 6

Measure average exec time of your statements with `timeit` module. As your submission, fill out the table with results (rounded to 2 decimal places)

In [11]:
import timeit

results = {
    "mean": (
        timeit.timeit(stmt_mean_custom, setup=setup, globals=globals(), number=10_000),
        timeit.timeit(stmt_mean_np, setup=setup, globals=globals(), number=10_000),
    ),
    "var": (
        timeit.timeit(stmt_var_custom, setup=setup, globals=globals(), number=10_000),
        timeit.timeit(stmt_var_np, setup=setup, globals=globals(), number=10_000),
    ),
    "std": (
        timeit.timeit(stmt_std_custom, setup=setup, globals=globals(), number=10_000),
        timeit.timeit(stmt_std_np, setup=setup, globals=globals(), number=10_000),
    ),
    "median": (
        timeit.timeit(stmt_median_custom, setup=setup, globals=globals(), number=10_000),
        timeit.timeit(stmt_median_np, setup=setup, globals=globals(), number=10_000),
    ),
}

print("Time per 10000 executions, secs")
print("Func\tCustom\tNumpy")
for func, times in results.items():
    print(f"{func}\t{times[0]:.2f}\t{times[1]:.2f}")


Time per 10000 executions, secs
Func	Custom	Numpy
mean	0.54	0.09
var	13.18	0.39
std	13.21	0.40
median	18.02	2.64


Time per 10000 executions, secs

| Func       | Custom | Numpy |
| ---------- | ------ | ----- |
| mean       |0.54    |0.09   |
| var        |13.18   |0.39   |
| std        |13.21   |0.40   |
| median     |18.02   |2.64   |